### DISCUSSION TITLE
company-backends vs. completion-at-point-functions

### DISCUSSION THREAD

**Discussion by clemera** [company-mode/company-mode#1173] [Q&A] [answered]
What kind of features can be achieved by company-backends which aren't possible by using `completion-at-point-functions` via `company-capf`? In which cases is it recommended for package authors to provide company backends  instead of providing functions for `completion-at-point-functions`?

Is there some plan to integrate backends into `completion-at-point-functions` someday via some kind of wrapper or would that be incompatible/impossible/undesirable?

**Comment by dgutov** [comment, marked as answer]
> What kind of features can be achieved by company-backends which aren't possible by using completion-at-point-functions via company-capf?

Async calling convention and backend grouping.

> In which cases is it recommended for package authors to provide company backends instead of providing functions for completion-at-point-functions?

I hesitate to make an official recommendation.

> Is there some plan to integrate backends into completion-at-point-functions someday via some kind of wrapper or would that be incompatible/impossible/undesirable?

There is some plan, but we'll have to wait and see. I'm hoping for a new version of the `completion-at-point-functions` API.

**Comment by clemera** [comment]
Thank you very much that is all I wanted to know, sometimes the Emacs completion landscape can be a bit confusing. Thank you for all your nice work on company, it's a great achievement!

**Comment by minad** [comment]
My [Cape package](https://github.com/minad/cape) provides an [adapter](https://github.com/minad/cape#company-adapter), which can convert a Company backend to a Capf. This allows you to use Company backends with frontends which rely on the Capf API only (Corfu, consult-completion-in-region, selectrum-completion-in-region, etc). The adapter is still experimental and has some limitations, in particular querying multiple asynchronous Company backends in parallel for grouping is not yet supported.
